### PR TITLE
feat(gateway): extend ICE timeout

### PR DESCRIPTION
**Pull request by thomaseizinger**
Currently, a `snownet` Client and Server always have the same ICE timeout configuration. This doesn't necessarily have to be the case. A Gateway cannot establish connections to a Client anyway and thus, we can have much laxer requirements on when we detect that a Client has disappeared (without saying "goodbye").

Extending the idle and default ICE timeout values should hopefully reduce the number of false-positive disconnects that users may experience where a Gateway cuts a connection because it believes the Client is gone when in reality, perhaps a few STUN packets just got lost or backed up.

Changing the ICE timeout exposes a few corner-cases in how we track and use time within `snownet`. In particular, it is now obviously possible for a Gateway to still retain the connection state of a Client whilst the Client has long disconnected but now reconnects using the same ICE credentials and private key.

Our proptests uncovered some state misalignment in that scenario due to some remaining time impurity within `boringtun` (see https://github.com/firezone/boringtun/pull/126 for details). In addition, our idle state transitions needed to be updated to also take into account candidate changes on both sides in order to achieve a deterministic outcome.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
- Align this with the WG timeout?
- Remove less useful proptest transitions

**Review by jamilbk — APPROVED** [review]
Giving a tentative approval here based on concept. I get a weird feeling there is some edge case or two that will pop up we didn't anticipate, but we can do some soak testing to see.

LGTM pending @klochowicz's review.

**Comment by thomaseizinger** [comment]
> Giving a tentative approval here based on concept. I get a weird feeling there is some edge case or two that will pop up we didn't anticipate, but we can do some soak testing to see.

Yeah, any changes are these timeouts are always a bit risky. The tests are covering this pretty well and with the additionally removed time-impurity in `boringtun`, I am feeling a bit better about this overall. Won't replace soak testing of course.

**Comment by thomaseizinger** [comment]
CI will be fixed as a result of #10907.

File: rust/connlib/snownet/src/node.rs [hunk 1]
```diff
@@ -138,22 +139,27 @@ pub struct Node<T, TId, RId> {
...
     buffer_pool: BufferPool<Vec<u8>>,
 
-    mode: T,
+    role: T,
     rng: StdRng,
+
+    /// The number of seconds since the UNIX epoch.
+    unix_ts: Duration,
+    /// The [`Instant`] at the time we read the UNIX epoch above.
+    unix_now: Instant,
```
Comment (klochowicz) [inline rust/connlib/snownet/src/node.rs:148]: do we need both of these fields? as `unit_ts` is unused in this PR, I can't tell it's purpose

**Review by klochowicz — APPROVED** [review]

File: rust/connlib/snownet/src/node.rs (hunk 1 above)
Comment (thomaseizinger) [reply rust/connlib/snownet/src/node.rs:148]: It is now unused, it is passed into `Tunn::new_at`.

File: rust/connlib/snownet/src/node.rs [hunk 2]
```diff
@@ -763,6 +774,8 @@ where
             Some(self.rate_limiter.clone()),
             self.rng.next_u64(),
             now,
+            self.unix_now,
+            self.unix_ts,
```
Comment (thomaseizinger) [inline rust/connlib/snownet/src/node.rs:741]: @klochowicz It is used here.

File: rust/connlib/snownet/src/node.rs (hunk 1 above)
Comment (thomaseizinger) [reply rust/connlib/snownet/src/node.rs:148]: We do need both, see https://github.com/firezone/boringtun/pull/126 for reference.

`unix_ts` are the number of seconds since the UNIX epoch. `unix_now` is the `Instant` at the same time we took `unix_ts`. The combination of the two allows us to compute the UNIX timestamp for a given `Instant`.
